File: params.go
package main

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

// paramPattern matches a single parameter value: numbers, pi expressions, or combinations.
// Examples: "1.5707", "pi", "pi/2", "3*pi/4", "-pi", "-2*pi/3", "3.14e-2"
const paramPattern = `-?(?:\d*\.?\d*\*?pi(?:/\d+\.?\d*)?|\d+\.?\d*(?:[eE][+\-]?\d+)?)`

// piExprRegex matches expressions like: pi, 2pi, 2*pi, pi/2, 3pi/4, 3*pi/4, -pi, -pi/2, -3*pi/4
var piExprRegex = regexp.MustCompile(`^(-?)(\d*\.?\d*)\s*\*?\s*pi(?:\s*/\s*(\d+\.?\d*))?$`)

// parseParamExpr parses a single parameter expression, supporting plain numbers and pi expressions.
// Returns the parsed float64 value and true on success, or 0 and false on failure.
//
// Supported formats:
//   - Plain numbers: "1.5707", "3.14", "-0.5"
//   - Pi constant: "pi"
//   - Pi fractions: "pi/2", "pi/4", "pi/3"
//   - Coefficients: "2pi", "2*pi", "3pi/4", "3*pi/4"
//   - Negative: "-pi", "-pi/2", "-3*pi/4"
func parseParamExpr(s string) (float64, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, false
	}

	// Try plain number first
	if val, err := strconv.ParseFloat(s, 64); err == nil {
		return val, true
	}

	// Try pi expression
	s = strings.ToLower(s)
	if matches := piExprRegex.FindStringSubmatch(s); matches != nil {
		negative := matches[1] == "-"
		coeffStr := matches[2]
		denomStr := matches[3]

		coeff := 1.0
		if coeffStr != "" {
			var err error
			coeff, err = strconv.ParseFloat(coeffStr, 64)
			if err != nil {
				return 0, false
			}
		}

		result := coeff * math.Pi

		if denomStr != "" {
			denom, err := strconv.ParseFloat(denomStr, 64)
			if err != nil || denom == 0 {
				return 0, false
			}
			result /= denom
		}

		if negative {
			result = -result
		}
		return result, true
	}

	return 0, false
}

// formatParam formats a float64 parameter value, using pi notation when possible.
// Recognizes common pi fractions: pi, pi/2, pi/4, pi/3, pi/6, pi/8, 2pi, 3pi/4, etc.
func formatParam(val float64) string {
	// Table of recognized pi fractions: coefficient, denominator, display string
	type piForm struct {
		value   float64
		display string
	}
	piForms := []piForm{
		{2 * math.Pi, "2*pi"},
		{math.Pi, "pi"},
		{math.Pi / 2, "pi/2"},
		{math.Pi / 3, "pi/3"},
		{math.Pi / 4, "pi/4"},
		{math.Pi / 6, "pi/6"},
		{math.Pi / 8, "pi/8"},
		{3 * math.Pi / 4, "3*pi/4"},
		{3 * math.Pi / 2, "3*pi/2"},
		{2 * math.Pi / 3, "2*pi/3"},
	}

	for _, pf := range piForms {
		if math.Abs(val-pf.value) < 1e-10 {
			return pf.display
		}
		if math.Abs(val+pf.value) < 1e-10 {
			return "-" + pf.display
		}
	}

	return fmt.Sprintf("%g", val)
}

// parseParams parses a parameter string into float values.
// Returns nil if any part fails to parse.
func (m *Model) parseParams(input string) []float64 {
	var params []float64
	parts := strings.Split(input, ",")
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		val, ok := parseParamExpr(part)
		if !ok {
			return nil // validation failure
		}
		params = append(params, val)
	}
	return params
}

File: src/params.rs
use regex::Regex;
use std::f64::consts::PI;
use std::sync::OnceLock;

static PI_EXPR_REGEX: OnceLock<Regex> = OnceLock::new();

pub fn parse_param_expr(s: &str) -> Option<f64> {
    let s = s.trim();
    if s.is_empty() {
        return None;
    }

    if let Ok(val) = s.parse::<f64>() {
        return Some(val);
    }

    let s = s.to_lowercase();
    let re = PI_EXPR_REGEX.get_or_init(|| {
        Regex::new(r"^(-?)(\d*\.?\d*)\s*\*?\s*pi(?:\s*/\s*(\d+\.?\d*))?$").unwrap()
    });

    if let Some(caps) = re.captures(&s) {
        let negative = &caps[1] == "-";
        let coeff_str = &caps[2];
        let denom_str = caps.get(3).map_or("", |m| m.as_str());

        let mut coeff = 1.0;
        if !coeff_str.is_empty() {
            coeff = coeff_str.parse::<f64>().ok()?;
        }

        let mut result = coeff * PI;

        if !denom_str.is_empty() {
            let denom = denom_str.parse::<f64>().ok()?;
            if denom == 0.0 {
                return None;
            }
            result /= denom;
        }

        if negative {
            result = -result;
        }
        return Some(result);
    }

    None
}

pub fn format_param(val: f64) -> String {
    struct PiForm {
        value: f64,
        display: &'static str,
    }

    let pi_forms = [
        PiForm {
            value: 2.0 * PI,
            display: "2*pi",
        },
        PiForm {
            value: PI,
            display: "pi",
        },
        PiForm {
            value: PI / 2.0,
            display: "pi/2",
        },
        PiForm {
            value: PI / 3.0,
            display: "pi/3",
        },
        PiForm {
            value: PI / 4.0,
            display: "pi/4",
        },
        PiForm {
            value: PI / 6.0,
            display: "pi/6",
        },
        PiForm {
            value: PI / 8.0,
            display: "pi/8",
        },
        PiForm {
            value: 3.0 * PI / 4.0,
            display: "3*pi/4",
        },
        PiForm {
            value: 3.0 * PI / 2.0,
            display: "3*pi/2",
        },
        PiForm {
            value: 2.0 * PI / 3.0,
            display: "2*pi/3",
        },
    ];

    for pf in &pi_forms {
        if (val - pf.value).abs() < 1e-10 {
            return pf.display.to_string();
        }
        if (val + pf.value).abs() < 1e-10 {
            return format!("-{}", pf.display);
        }
    }

    val.to_string()
}

pub fn parse_params(input: &str) -> Option<Vec<f64>> {
    let mut params = Vec::new();
    for part in input.split(',') {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }
        if let Some(val) = parse_param_expr(part) {
            params.push(val);
        } else {
            return None; // validation failure
        }
    }
    Some(params)
}
